### DISCUSSION TITLE
Incrementing variables

### DISCUSSION THREAD

**Discussion by wordmuse** [freeplane/freeplane#2100] [Ideas]
I was wondering if it might be possible for Freeplane to support incrementing variables. 

Some possible use cases: 

- journaling, where the journal entry number increments with each entry
- page numbers and chapter numbers in a book or report
- footnote and endnote numbers
- items being catalogued
- experimental trial numbers

I don't know how difficult this would be to implement, but I thought it might be worth suggesting.

If this is already available, that would be super. 

~ Bal

**Comment by Coffee-Search** [comment]
This is possible with scripting / add ons.

I am new to the freeplane API but the following is a very quick, simple, example to demonstrate for creating new ticket nodes where the counter (incrementing variable) and prefix are stored against the root node as attributes. 

- Tools > open user directory
- save the code to the directory _scripts_ in a file named NewTicket.groovy 
- close/open freeplane
- _New Ticket_ Should now be available in _Tools > scripts > New Ticket_ and _right click on node > scripts  > New Node_ . You can also bind it to a hot key o call the script using _Tools > Assign hot key_

Note, attributes can be hidden using ...
View > Node Attributes > Hide all attributes
View > Node Attributes > Show icon for attributes
Or individually for a selected node Alt+F3 (View > Minimize Node)

This is just a simple example and could contain more defensive code / error checking, etc. There are other places/options to store the variable, etc. and I am sure there is functionality to provide better integration depending on your requirement. 

```
// @ExecutionModes({ON_SINGLE_NODE}) 

COUNTER_ATTR   = 'ticketCounter'
PREFIX_ATTR    = 'ticketPrefix'
REFERENCE_ATTR = 'ticketReference'

rootNode = node.mindMap.root

def getPrefix() {
  def value = rootNode[PREFIX_ATTR].text // gets calculated value of attribute
  if(value == null) {
    value = '#'
    rootNode.attributes.add([PREFIX_ATTR,value])
  }
  return(value)
}
def getCounter() {
  def value = rootNode.attributes.get(COUNTER_ATTR) // gets raw value of attribute
  if(value == null) {
    value = 1
    rootNode.attributes.add([COUNTER_ATTR,value])
  }
  rtn = value
  rootNode.attributes.set(COUNTER_ATTR,value+1)
  return(rtn)
}

def counter = getCounter()
def prefix  = getPrefix()
def reference = "$prefix$counter"
def newNode = node.createChild("[$reference] ")
newNode.attributes.add([REFERENCE_ATTR, reference])
c.edit(newNode)
```
UPDATE: added c.edit(newNode) to enter inline edit for node after creation
UPDATE: changed the fetching of the prefix so that a formula can be used, e.g.  **`= new java.util.Date().format( "yyyyMMdd-")`**

**Reply by euu2021** [reply]
Interesting!

Maybe, it's also possible to use the map storage, in order to make the counter more "hidden". [Example of usage](https://docs.freeplane.org/scripting/Scripts_collection.html#using-map-specific-storage) of the map storage.

**Reply by wordmuse** [reply]
That is a very interesting script. Three questions. Is there a way to reset the count?  Is there a way to make it start over for a given attribute, e.g., chapter number, page number, citation number, etc.? Is there a way to manually edit the count and have the incrementing script start where I specify? For example, if I want to skip numbers 8-20, knowing that I will be manually entering them at a later time, have it skip from 7 to 21?

Thanks!

~ Bal

**Reply by Coffee-Search** [reply]
> Maybe, it's also possible to use the map storage, in order to make the counter more "hidden".

Yes I had this thought - its probably a good idea to. It would stop manual tampering and features like reset can be implemented using scripts if need. 

> Is there a way to reset the count?

Yes in this example simply edit the ticketCounter attribute on the root node or write a script called ResetTicketCounter.groovy (or similar) to do the same. That script could also modify the prefix based on date, etc. I am sure there must be a way to use a formula for the prefix, e.g. `= new java.util.Date().format( "MMM-yyyy-")`  and for a script to read the calculated value not the formula text - not found the API reference for that yet.
> Is there a way to make it start over for a given attribute, e.g., chapter number, page number, citation number, etc.?

If I understand the question yes. Beyond the given example you can walk back from where you are in the map for the nearest Chapter node. In this way you can have a Chapter node with all he attributes you need on it that could be set up with a script called NewChapter.groovy. Other scripts like NewCitation.groovy, NewSubSection.groovy, etc. can look back for the nearest Chapter node and use the attributes from it accordingly.  The root node can be used to hold the next chapter number, etc. You could also utilise depth as a reference.
> Is there a way to manually edit the count and have the incrementing script start where I specify?

Yes you can specify an attribute something like preAllocatedNumbers and have a comma separated list of numbers and ranged that the script can read and avoid/skip when incrementing the number counter, e.g. "5,500,8-20,2020". Following my ticket example I would also set up a script to create a blank ticket, e.g. NewManualTicket.groovy, that would create a blank node with all the attributes you wanted, e.g. TicketReference, to be left for you to fill in manually. 

For things like ticket, requirement, feature, case, project, product, asset, etc. numbering you only need a unique reference that can be moved around a map with out changing the reference. The ticket example is good for these types of allocation. I am all for atomising documentation so that a unique reference can be maintained. 

For publishing, contracts and documents it is more about the logical position in the flow of the map. The current scheme used by freeplane (Node numbering) can do this but I have not found a way to reference the number displayed/allocated which would be necessary to add/group annotations like citations, etc. I would look to 'type' the node as 'chapter header', 'sub heading', 'citation', etc. and make a scripts/addon work on that basis. I might be possible for you could include tags in the body of node text that a script could pick up and use to automatically update, e.g. "#subheading" can be replaced with "1.2.4 \<!--#subheading--\>". You would have to consider what to do about reordering/moving chapters, etc.  I think the ticket example demonstrates the features are probably there to do what you want. To communicate your requirement further it might be best to mock up a map.
